src/lib.rs

```diff
@@ -595,6 +596,237 @@ impl<A: Array> ArrayVec<A> {
+         where A::Item: Clone,
+    {
+        let len = self.len();
+        if new_len > len {
+            self.extend_with(ExtendElement::new(new_len - len, value))
+        } else {
+            self.truncate(new_len);
+            Ok(())
+        }
+    }
+
+    /// Resizes the `ArrayVec` in-place so that its length becomes `new_len`.
+    ///
+    /// Does the same thing as `try_resize_default`, but panics on error
+    /// instead of returning it as a `Result`.
+    ///
+    /// **Panics** if the corresponding `try_resize_default` would return an
+    /// error.
+    pub fn resize_default(&mut self, new_len: usize) where A::Item: Default {
+        self.try_resize_default(new_len).unwrap();
+    }
+
+    /// Resizes the `ArrayVec` in-place so that its length becomes `new_len`.
+    ///
+    /// If `new_len` is greater than the vector's current length, the
+    /// `ArrayVec` is extended by the difference, with each additional slot
+    /// filled with `Default::default()`. If `new_len` is less than the
+    /// vector's current length, the `ArrayVec` is simply truncated.
+    ///
+    /// This method uses `Default` to create new values on every push. If you'd
+    /// rather `Clone` a given value, use `try_resize`.
+    ///
+    /// If the capacity of the `ArrayVec` is smaller than the passed `new_len`,
+    /// an error is returned, and no changes to the `ArrayVec` is made.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use arrayvec::ArrayVec;
+    ///
+    /// let mut vec: ArrayVec<[_; 8]> = (1..4).collect();
+    /// assert!(vec.try_resize_default(5).is_ok());
+    /// assert_eq!(&vec[..], &[1, 2, 3, 0, 0]);
+    ///
+    /// let mut vec: ArrayVec<[_; 4]> = (1..5).collect();
+    /// assert!(vec.try_resize_default(2).is_ok());
+    /// assert_eq!(&vec[..], &[1, 2]);
+    /// ```
+    pub fn try_resize_default(&mut self, new_len: usize)
+        -> Result<(), CapacityError>
+        where A::Item: Default,
+    {
+        let len = self.len();
+        if new_len > len {
+            self.extend_with(ExtendDefault::new(new_len - len))
+        } else {
+            self.truncate(new_len);
+            Ok(())
+        }
+    }
+}
+
+trait ExtendIterator: Iterator {
+    fn needed_capacity(&self) -> usize;
+}
+
+struct ExtendElement<T: Clone>(usize, Option<T>);
+impl<T: Clone> ExtendElement<T> {
+    fn new(n: usize, value: T) -> ExtendElement<T> {
+        ExtendElement(n, Some(value))
+    }
+}
+impl<T: Clone> Iterator for ExtendElement<T> {
+    type Item = T;
+    fn next(&mut self) -> Option<T> {
+        if self.0 == 0 {
+            return None;
+        }
+        self.0 -= 1;
+        Some(if self.0 > 0 {
+            self.1.as_ref().unwrap().clone()
+        } else {
+            self.1.take().unwrap()
+        })
+    }
+}
+impl<T: Clone> ExtendIterator for ExtendElement<T> {
+    fn needed_capacity(&self) -> usize {
+        self.0
+    }
+}
+
+struct ExtendDefault<T: Default>(usize, PhantomData<T>);
```

Comment:
The ExtendDefault iterator is not needed, it can be passed as a regular iterator inline, maybe something like `(0..n).map(|_| X::default())`